fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -140,6 +155,106 @@ instance MonadDB (Transaction Server) Server where
 instance MonadRelayStore Server where
   getStoreVar = asks linkRelayStoreVar
 
+instance MonadPowerDNS Server where
+  set recType url zoneTxt contents ttl = do
+    logDebug $ mconcat
+      [ "Set DNS "
+      , displayShow recType
+      , " Record "
+      , displayShow url
+      , " to "
+      , displayShow contents
+      ]
+    clientEnv <- pdnsClientEnv
+    let
+      rrset =
+        RRSet { rrset_name = Text.append (textDisplay url) "."
+              , rrset_type = recType
+              , rrset_ttl = ttl
+              , rrset_changetype = Just Replace
+              , rrset_records = Just $ formatRecords contents
+              , rrset_comments = Nothing
+              }
+      rrsets = RRSets [rrset]
+    resp <- liftIO $ runClientM (PDNS.updateRecords "localhost" zoneTxt rrsets) clientEnv
+    case resp of
+      Left err -> do
+        logWarn @Text "PowerDNS.set failed"
+        return . Left $ Web.Error.toServerError err
+      Right good -> do
+        logDebug @Text "PowerDNS.set succeeded"
+        return $ Right good
+    where
+      formatRecords :: NonEmpty Text -> [Record]
+      formatRecords values =
+        fmap newRecord (toList values)
+
+      newRecord :: Text -> Record
+      newRecord content =
+        Record { record_content = format recType content, record_disabled = False }
+
+      format :: PowerDNS.API.Zones.RecordType -> Text -> Text
+      format TXT = UTF8.wrapIn "\""
+      format _   = identity
+
+  get url _ = do
+    logDebug $ "PowerDNS.get for " <> textDisplay url
+    clientEnv <- pdnsClientEnv
+    resp <- liftIO $ runClientM (PDNS.search "localhost" (textDisplay url) 1 (Just TyRecord)) clientEnv
+    case resp of
+      Left err -> do
+        logWarn @Text "PowerDNS.get failed"
+        return . Left $ Web.Error.toServerError err
+      Right good -> do
+        logDebug @Text "PowerDNS.get succeeded"
+        return $ Right good
+
+  clear url zoneID = do
+    logDebug $ "PowerDNS.clear " <> textDisplay url
+    results <- PowerDNS.get url zoneID
+    case results of
+      Left err -> do
+        return $ Left err
+      Right searchResults -> do
+        mapM_ removeResult searchResults
+        return $ Right ()
+    where
+      removeResult :: (MonadIO m, MonadReader Config m, MonadThrow m) => PDNS.SearchResult -> m ()
```

Comment:
Amazing that you're diving right into abstraction! 👏 

No need to change anything, but we're also working directly in `Server a`, which has instances for all of these. So _in this context,_ we don't need to generalize:

```haskell
removeResult :: Server ()
```

A lot of our other code is pretty general, because we want to be able to plug in other contexts, such as test environments that stub out DNS access and whatnot. Actually with this PDNS stuff, we can possibly think about making a lot of the code more concrete, and "just" using `Server a` everywhere since we can spin up DNS locally 🤔🤔🤔🤔🤔🤔🤔

(It _is_ nice to have the ability to easily stub out anything in a super clean, first class way, but also it has typeclass overhead, so worth thinking about after this PR makes it possible)

Reply:
> removeResult :: Server ()

Let me know if why you can do this here is confusing, and we can chat about it 👍